### PR TITLE
Get/set tta

### PR DESCRIPTION
**Pull request by hell0w0rld2**
Changes the turnToAngle class from what it was before to a working state machine.

### CODE REVIEW

**Review by ManiacMoustache — APPROVED** [review]
+1

**Review by ManiacMoustache — DISMISSED** [review]
+1

File: components/turnToAngle.py [hunk 1]
```diff
@@ -1,101 +1,97 @@
...
-    @feedback
-    def nextHeadingDisplay(self):
-        return self.nextHeading
+    @state(first = True)
+    def idling(self):
+        """Stays in this state until started"""
+        if self.turnAngle != 0:
+            self.next_state("turn")
+        else:
+            log.error("Must have an angle to turn to")
```
Comment (tuffinmuffin) [inline components/turnToAngle.py:43]: Do you really want this this to spam?

File: components/turnToAngle.py [hunk 2]
```diff
@@ -1,101 +1,97 @@
...
+        """Turns the robot based off of the speed determined in setSpeedFunc"""
+        self.setSpeedFunc()
+        if self.change > 0:
+            self.driveTrain.setTank(-1 * self.speed, self.speed)
+        else:
+            self.driveTrain.setTank(self.speed, -1 * self.speed)
+        self.next_state("turn")
+
+        """Stops the automatic turning if the bot is within the tolerance of the desired angle"""
+        if self.navx.getFusedHeading() <= self.nextHeading + self.tolerance and self.navx.getFusedHeading() >= self.nextHeading - self.tolerance:
```
Comment (tuffinmuffin) [inline components/turnToAngle.py:85]: Can do a abs(getFusedHeading-nextHeading) < tolerance

File: components/turnToAngle.py [hunk 3]
```diff
@@ -1,101 +1,97 @@
...
+            self.driveTrain.setTank(0, 0)
+            self.stop()
+
+    def stop(self):
+        self.running = False
+        self.starting = False
+        self.initialHeading = self.navx.getFusedHeading()
```
Comment (tuffinmuffin) [inline components/turnToAngle.py:92]: Why get the fused heading here? Seems like just need to get it when a new angle is commanded? Might not be the same when we start moving again.

File: robot.py [hunk 4]
```diff
@@ -160,13 +163,14 @@ def teleopPeriodic(self):
...
             executingDriveCommand = True
             self.autoAlign.setShootAfterComplete(True)
             self.autoAlign.engage()
         if self.xboxMap.getDriveX() == True:
             executingDriveCommand = True
-            self.turnToAngle.setIsRunning()
+            self.turnToAngle.start()
         else:
             self.turnToAngle.stop()
         if self.xboxMap.getDriveA() == False and self.prevAState == True:
```
Comment (tuffinmuffin) [inline robot.py:176]: 5 Stars would order again.

Comment (tuffinmuffin) [reply robot.py:176]: Order arrived broken. 5 stars

Comment (tuffinmuffin) [reply robot.py:176]: Does the work with a toothbrush? I can't find it in the product listing.

File: components/turnToAngle.py [hunk 5]
```diff
@@ -1,101 +1,97 @@
...
+            self.change += 360
 
-    def execute(self):
-        self.output()
-        self.PIDController = controller.PIDController(Kp= self.P, Ki= self.I, Kd= self.D, period = self.time)
-        self.heading = self.navx.getFusedHeading()
+    def setSpeedFunc(self):
+        """Determines the speed based off of the distance from the desired angle"""
+        self.calcHeading()
+        if abs(self.change) > 90:
```
Comment (tuffinmuffin) [inline components/turnToAngle.py:67]: Could simplify from 3 if elsif to a if/elseif/else or to 2 if() that change from a default (speed = speed for > 90 if(90 <) (set < 90 speed); if(<slow) set slow;

File: components/turnToAngle.py (hunk 3 above)
Comment (tuffinmuffin) [reply components/turnToAngle.py:92]: Shoudn't this set the angle goal to 0 to stop turning?

**Review by ManiacMoustache — APPROVED** [review]
+1
Haha statemachine go brrrrrrr

**Review by Leaf-life — APPROVED** [review]
itll probly work

**Pull request merged**